File: scan.java
import java.io.FileInputStream;
import java.io.IOException;
import java.util.List;

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
import com.google.api.services.sheets.v4.Sheets;
import com.google.api.services.sheets.v4.SheetsScopes;
import com.google.api.services.sheets.v4.model.Spreadsheet;
import com.google.api.services.sheets.v4.model.ValueRange;
import com.google.zxing.BinaryBitmap;
import com.google.zxing.DecodeHintType;
import com.google.zxing.MultiFormatReader;
import com.google.zxing.NotFoundException;
import com.google.zxing.Reader;
import com.google.zxing.Result;
import com.google.zxing.common.HybridBinarizer;
import com.google.zxing.client.j2se.BufferedImageLuminanceSource;
import com.google.zxing.client.j2se.ImageReader;

import javax.sound.sampled.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.util.Collections;

public class BarcodeScannerApp {

    // Path to the Google Sheets credentials JSON file
    private static final String CREDENTIALS_FILE = "scan-390611-25fc1daa0744.json";

    // Google Sheets document ID
    private static final String DOCUMENT_ID = "1kTClIxeAcZhwXufiGUhxbJTn5X24fhIcjsxNS9Do9Ew";

    // Function to append barcode data to Google Sheets and validate against sheet data
    private static void appendToSheet(String barcodeData) throws IOException {
        // Create a Google credential from the service account credentials JSON file
        GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(CREDENTIALS_FILE))
                .createScoped(Collections.singleton(SheetsScopes.SPREADSHEETS));

        // Create a Google Sheets service
        Sheets sheetsService = new Sheets.Builder(credential.getTransport(), credential.getJsonFactory(), credential)
                .setApplicationName("Barcode Scanner App")
                .build();

        // Open Sheet 2 within the Google Sheet document
        String sheetName = "Sheet2"; // Replace with the name of Sheet 2

        // Get all values from Sheet 2
        ValueRange response = sheetsService.spreadsheets().values()
                .get(DOCUMENT_ID, sheetName + "!A1:C")
                .execute();

        List<List<Object>> sheetValues = response.getValues();

        // Search for the scanned barcode data in Sheet 2
        for (List<Object> row : sheetValues) {
            if (row.get(0).toString().equals(barcodeData)) {
                // Print the corresponding 3PL name and company name
                if (row.size() > 1) {
                    System.out.println("3PL Name: " + row.get(1));
                }
                if (row.size() > 2) {
                    String companyName = row.get(2).toString();
                    System.out.println("Company Name: " + companyName);

                    // Play sound effect based on company name
                    playSoundEffect(companyName);
                }

                // Check for duplicate entry in Sheet 1
                ValueRange existingData = sheetsService.spreadsheets().values()
                        .get(DOCUMENT_ID, "Sheet1!A:A")
                        .execute();

                List<List<Object>> existingValues = existingData.getValues();

                boolean isDuplicate = false;
                for (List<Object> rowData : existingValues) {
                    if (rowData.get(0).toString().equals(barcodeData)) {
                        isDuplicate = true;
                        break;
                    }
                }

                if (isDuplicate) {
                    System.out.println("Error: Duplicate entry.");
                    // Play sound effect for duplicate entry
                    playSoundEffect("error");
                } else {
                    // Append the barcode data to Sheet 1
                    sheetsService.spreadsheets().values()
                            .append(DOCUMENT_ID, "Sheet1", new ValueRange().setValues(Collections.singletonList(Collections.singletonList(barcodeData))))
                            .setValueInputOption("RAW")
                            .execute();

                    System.out.println("Data saved successfully.");
                }
                break;
            }
        }
        // Scanned barcode data not found in Sheet 2 or did not validate
        System.out.println("Error: Barcode not found or data did not validate.");
        // Play sound effect for invalid data
        playSoundEffect("error");
    }

    // Barcode scanning function
    private static void scanBarcodes() throws IOException, NotFoundException {
        // Initialize the camera or capture device

        // Initialize the barcode reader
        Reader reader = new MultiFormatReader();

        while (true) {
            // Capture frame or image

            // Decode barcodes
            BufferedImage bufferedImage = null; // Replace with the captured frame or image

            BinaryBitmap binaryBitmap = new BinaryBitmap(new HybridBinarizer(new BufferedImageLuminanceSource(bufferedImage)));
            Result[] barcodes = reader.decodeMultiple(binaryBitmap);

            for (Result barcode : barcodes) {
                // Extract barcode data
                String barcodeData = barcode.getText();
                System.out.println("Scanned Barcode: " + barcodeData);
                appendToSheet(barcodeData);
            }

            // Display the frame or image

            // Wait for 'q' key to exit
            // if (key == 'q') break;
        }

        // Release the capture device or camera
        // Close the window or UI component
    }

    // Manual barcode entry function
    private static void manualEntry() throws IOException {
        while (true) {
            // Read barcode data from user input

            // if (barcodeData.equals("q")) break;

            appendToSheet(barcodeData);
        }
    }

    // Play sound effect based on company name
    private static void playSoundEffect(String companyName) {
        try {
            String soundFileName;
            if (companyName.equals("Nayra")) {
                soundFileName = "Nayra.wav";
            } else if (companyName.equals("Ni")) {
                soundFileName = "Ni.wav";
            } else if (companyName.equals("DH")) {
                soundFileName = "DH.wav";
            } else {
                soundFileName = "error.wav";
            }

            File soundFile = new File(soundFileName);
            AudioInputStream audioInputStream = AudioSystem.getAudioInputStream(soundFile);
            Clip clip = AudioSystem.getClip();
            clip.open(audioInputStream);
            clip.start();
        } catch (UnsupportedAudioFileException | IOException | LineUnavailableException e) {
            e.printStackTrace();
        }
    }

    // Main function
    public static void main(String[] args) {
        System.out.println("Barcode Scanner & Data Entry App");
        System.out.println("1. Scan Barcodes");
        System.out.println("2. Manual Entry");

        // Read user's choice

        if (choice.equals("1")) {
            try {
                scanBarcodes();
            } catch (IOException | NotFoundException e) {
                e.printStackTrace();
            }
        } else if (choice.equals("2")) {
            try {
                manualEntry();
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            System.out.println("Invalid choice. Exiting...");
        }
    }
}

File: scan.py
import os
import tkinter as tk
from googleapiclient.discovery import build
from google.oauth2 import service_account
import winsound

# Google Sheets API credentials
SCOPES = ['https://www.googleapis.com/auth/spreadsheets']
CREDS_FILE = 'scan-390611-25fc1daa0744.json'  # Path to your credentials JSON file
SHEET_ID = '1JvvmA_Lbr55XwYD2atYzeoAbD_ntOJxPx9IDKyCbK5c'  # ID of the Google Sheet

# Create GUI
root = tk.Tk()
root.title('Ni Fashion')

# Entry field
lbl_awb = tk.Label(root, text='AWB:')
lbl_awb.grid(row=0, column=0)
entry_awb = tk.Entry(root)
entry_awb.grid(row=0, column=1)

# Labels for displaying 3PL and company names
lbl_3pl = tk.Label(root, text='3PL:')
lbl_3pl.grid(row=1, column=0)
lbl_company = tk.Label(root, text='Company:')
lbl_company.grid(row=2, column=0)

# Label for displaying error or duplicate message
lbl_message = tk.Label(root, text='')
lbl_message.grid(row=3, columnspan=2)

# Google Sheets API authentication
creds = service_account.Credentials.from_service_account_file(CREDS_FILE, scopes=SCOPES)

# Connect to Google Sheets API
service = build('sheets', 'v4', credentials=creds)

# Function to submit data
def submit_data(event=None):
    awb = entry_awb.get()

    # Get data from Data to check for duplicates
    data_range = 'Data!A:A'
    result = service.spreadsheets().values().get(
        spreadsheetId=SHEET_ID,
        range=data_range
    ).execute()
    data_values = result.get('values', [])

    # Check for duplicates in Data sheet
    duplicate = False
    for row in data_values:
        if len(row) > 0 and awb == row[0]:
            duplicate = True
            lbl_message.config(text='Duplicate AWB!')
            # Play duplicate validation sound
            winsound.PlaySound('sound3.wav', winsound.SND_FILENAME)
            break

    if not duplicate:
        # Get data from Manifest sheet for validation and corresponding names
        validation_range = 'Manifest!D:F'
        result = service.spreadsheets().values().get(
            spreadsheetId=SHEET_ID,
            range=validation_range
        ).execute()
        validation_data = result.get('values', [])

        # Perform validation against Manifest sheet data and get corresponding names
        valid = False
        for row in validation_data:
            if len(row) > 2 and awb == row[0]:
                valid = True
                lbl_3pl.config(text='3PL: ' + row[2])  # Display 3PL name
                lbl_company.config(text='Company: ' + row[1])  # Display company name
                break

        if valid:
            # Append data to Data sheet
            values = [awb]
            body = {'values': [values]}
            service.spreadsheets().values().append(
                spreadsheetId=SHEET_ID,
                range='Data!A:A',
                valueInputOption='USER_ENTERED',
                insertDataOption='INSERT_ROWS',
                body=body
            ).execute()
            # Play true validation sound
            winsound.PlaySound('sound1.wav', winsound.SND_FILENAME)
            lbl_message.config(text='Data added successfully!')
        else:
            lbl_message.config(text='Invalid AWB!')
            # Play false validation sound
            winsound.PlaySound('sound2.wav', winsound.SND_FILENAME)

    # Clear entry field after submitting
    entry_awb.delete(0, tk.END)

# Submit button
btn_submit = tk.Button(root, text='Submit', command=submit_data)
btn_submit.grid(row=0, column=2)

# Bind Enter key to submit data
root.bind('<Return>', submit_data)

# Run the GUI
root.mainloop()
